src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/OpenDocumentGenerator.cs

```diff
@@ -0,0 +1,182 @@
+                if (!_documentManager.TryGetDocument(documentSnapshot.Uri, out var newDocumentSnapshot) ||
+                    !newDocumentSnapshot.TryGetAllVirtualDocuments<CSharpVirtualDocumentSnapshot>(out virtualDocuments))
+                {
+                    // This should never happen.
+                    // The server clearly wants to tell us about a document in a project, but we don't know which project it's in.
+                    // Sadly there isn't anything we can do here to, we're just in a state where the server and client are out of
+                    // sync with their understanding of the document contents, and since changes come in as a list of changes,
+                    // the user experience is broken. All we can do is hope the user closes and re-opens the document.
+                    Debug.Fail($"Server wants to update {documentSnapshot.Uri} in {projectKey} but we don't know about the document being in any projects");
+                    _logger.LogError("[Cohost] Server wants to update {hostDocumentUri} in {projectKeyId} by we only know about that document in misc files. Server and client are now out of sync.", documentSnapshot.Uri, projectKey);
+                    return null;
+                }
+            }
+
+            foreach (var virtualDocument in virtualDocuments)
+            {
+                if (virtualDocument.ProjectKey.Equals(projectKey))
+                {
+                    _logger.LogDebug("[Cohost] Found C# virtual doc for {version}: {uri}", version, virtualDocument.Uri);
+
+                    return virtualDocument;
+                }
+            }
+
+            if (virtualDocuments.Length > 1)
+            {
+                // If the particular document supports multiple virtual documents, we don't want to try to update a single one
+                // TODO: Remove this eventually, as it is a possibly valid state (see comment below) but this assert will help track
+                // down bugs for now.
+                Debug.Fail("Multiple virtual documents seem to be supported, but none were updated, which is not impossible, but surprising.");
+            }
+
+            _logger.LogDebug("[Cohost] Couldn't find any virtual docs for {version} of {uri}", version, documentSnapshot.Uri);
+
+            // Don't know about document, no-op. This can happen if the language server found a project.razor.bin from an old build
+            // and is sending us updates.
+        }
+
+        return null;
+    }
+
+    private static HtmlVirtualDocumentSnapshot? TryGetHtmlSnapshot(LSPDocumentSnapshot documentSnapshot)
+    {
+        _ = documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlVirtualDocumentSnapshot);
+        return htmlVirtualDocumentSnapshot;
+    }
+
+    public void Initialize(ProjectSnapshotManagerBase projectManager)
+    {
+        _projectManager = projectManager;
+        _projectManager.Changed += ProjectManager_Changed;
```

Comment:
>ProjectManager_Changed [](http://example.com/codeflow?start=35&length=22)

I'm guessing we don't need to unsubscribe because lifecycle matches VS instance lifecycle?

Reply:
Correct. Also this whole file should be deleted in time :)